### PR TITLE
Add warnings to conversion, Warn for tuples size > 5

### PR DESCRIPTION
**Pull request by dylant-da**
Corresponding changes in ghc: https://github.com/digital-asset/ghc/pull/134

### CODE REVIEW

**Comment by basvangijzel-DA** [comment]
Still seems to be some failing tests.

File: compiler/damlc/lib/DA/Cli/Options.hs [hunk 1]
```diff
@@ -223,6 +223,12 @@ enableScenariosOpt = EnableScenarios <$>
             "Enable/disable support for scenarios as a language feature. \
             \If disabled, defining top-level scenarios is a compile-time error"
 
+allowLargeTuplesOpt :: Parser AllowLargeTuples
+allowLargeTuplesOpt = AllowLargeTuples <$>
+    flagYesNoAuto "allow-large-tuples" False desc internal
```
Comment (basvangijzel-DA) [inline compiler/damlc/lib/DA/Cli/Options.hs:228]: maybe no-warn-large-tuples or disable-warn-large-tuples?

Technically large tuples are already allowed, they just get a warning.

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs [hunk 2]
```diff
@@ -1967,6 +1975,17 @@ splitConArgs_maybe con args = do
 -- work. Constructor workers are not handled (yet).
 convertDataCon :: Env -> GHC.Module -> DataCon -> [LArg Var] -> ConvertM (LF.Expr, [LArg Var])
 convertDataCon env m con args
+    | AllowLargeTuples False <- envAllowLargeTuples env
+    , NameIn GHC_Tuple fsName <- con
+    , let name = unpackFS fsName
+    -- TODO: A better way to detect uses of tuples than this string malarkey
+    , let (front, middle, end) = (head name, init (tail name), last name)
```
Comment (basvangijzel-DA) [inline compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs:1982]: Is there any way of doing this more nicely with a new/existing pattern synonym?

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion/ConvertM.hs [hunk 3]
```diff
@@ -89,6 +94,21 @@ conversionError msg = do
...
+        { _range = maybe noRange sourceLocToRange convRange
+        , _severity = Just DsWarning
+        , _source = Just "Core to Daml-LF"
+        , _message = T.pack msg
+        , _code = Nothing
+        , _relatedInformation = Nothing
+        , _tags = Nothing
+        }
+      fileDiagnostic = (convModuleFilePath, ShowDiag, diagnostic)
+  modify $ \s -> s { warnings = warnings s ++ [fileDiagnostic] }
```
[the remark this answers is not in the document]
Comment (dylant-da) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion/ConvertM.hs:110]: Well spotted - I remember thinking "gotta add a TODO" when I typed that :sweat_smile: 

just fixed, I'll tackle the rest tomorrow morning

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs (hunk 2 above)
Comment (dylant-da) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs:1982]: Unfortunately, the data constructors for tuples are just basic data constructors at this stage, so I can't disambiguate them easily except through checking the strings and the point of import

File: compiler/damlc/daml-stdlib-src/DA/Internal/Desugar.daml [hunk 4]
```diff
@@ -142,3 +143,6 @@ newtype InterfaceView p i t = InterfaceView ()
 -- it's extracted unmodified.
 mkInterfaceView : forall p i t v. (Implements t i, HasInterfaceView i v) => (t -> v) -> InterfaceView p i t
 mkInterfaceView = magic @"mkInterfaceView"
+
+codeGenAllowLargeTuples : forall a. a -> a
```
[the remark this answers is not in the document]
Comment (dylant-da) [reply compiler/damlc/daml-stdlib-src/DA/Internal/Desugar.daml:147]: This allows the codegen to create tuples larger than size 5 without it triggering the size check in convertExpr

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion/ConvertM.hs [hunk 5]
```diff
@@ -55,11 +57,14 @@ newtype ConvertM a = ConvertM (ReaderT ConversionEnv (StateT ConversionState (Ex
 instance MonadFail ConvertM where
     fail = conversionError
 
-runConvertM :: ConversionEnv -> ConvertM a -> Either FileDiagnostic a
-runConvertM s (ConvertM a) = runExcept (evalStateT (runReaderT a s) st0)
+runConvertM :: ConversionEnv -> ConvertM a -> Either FileDiagnostic (a, [FileDiagnostic])
```
[the remark this answers is not in the document]
Comment (dylant-da) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion/ConvertM.hs:60]: Sure, I'll copy over some tests I wrote locally while making the GHC changes

[the remark this answers is not in the document]
Comment (dylant-da) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion/ConvertM.hs:62]: comment added

File: compiler/damlc/lib/DA/Cli/Options.hs (hunk 1 above)
Comment (dylant-da) [reply compiler/damlc/lib/DA/Cli/Options.hs]: Fixed

File: compiler/damlc/pkg-db/util.bzl [hunk 6]
```diff
@@ -120,6 +120,7 @@ def _daml_package_rule_impl(ctx):
         --target {daml_lf_version} \
         --cpp {cpp} \
         --ghc-option=-Werror \
+        --allow-large-tuples=yes \
```
[the remark this answers is not in the document]
Comment (dylant-da) [reply compiler/damlc/pkg-db/util.bzl]: Good catch, this was a stopgap measure before I added codeGenAllowLargeTuples to the GHC side
I've made this non-default, and we now get it from ctx

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion/ConvertM.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (dylant-da) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion/ConvertM.hs:62]: done

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs (hunk 2 above)
Comment (dylant-da) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs]: I've defined a pattern synonym which at least cleans up the code in LFConversion and offloads it to UtilGHC

**Review by basvangijzel-DA — APPROVED** [review]
Generally happy, what I didn't fully get is the code generation part. Please check with Remy/Moisés..

**Pull request merged**